File: 24_高级数据结构/good java/MinPairContainer.java
package template.primitve.generated.datastructure;

/**
 * maintain a0, ..., an and b0, ... , bn, supporting follow operations:
 * <ul>
 * <li>add new element a[n + 1] and b[n + 1] (a[n + 1] should be the largest element in a)</li>
 * <li>given x, for all i, set ai = min(ai, x)</li>
 * <li>find max(a[i] + b[i])</li>
 * </ul>
 * <p>
 * Each operation done in O(1) in average
 */
public class MinPairContainer {
  long[] as;
  long[] bs;
  int size = 0;

  public static final long inf = (long) 4e18;

  public MinPairContainer(int n) {
    as = new long[n];
    bs = new long[n];
  }

  public void clear() {
    size = 0;
  }

  private void insert(long a, long b) {
    as[size] = a;
    bs[size] = b;
    size++;
  }

  private void pop() {
    size--;
    // if (size > 0) {
    // cs[size - 1] = Math.min(cs[size - 1], cs[size]);
    // as[size - 1] = Math.min(as[size - 1], cs[size - 1]);
    // }
  }

  private long event(long a0, long b0, long b1) {
    // min(a0) + b0 >= min(a1) + b1
    // a1 >= a0
    if (b0 <= b1) {
      return -inf;
    }
    // a0 + b0 >= delta + b1
    return a0 + b0 - b1;
  }

  private void pushDown() {
    as[size - 2] = Math.min(as[size - 2], as[size - 1]);
  }

  public void add(long a, long b) {
    assert size == 0 || a >= as[size - 1];
    while (size > 1) {
      pushDown();
      long e1 = event(as[size - 2], bs[size - 2], bs[size - 1]);
      long e2 = event(as[size - 1], bs[size - 1], b);
      if (e1 >= e2) {
        pop();
      } else {
        break;
      }
    }
    if (size == 0 || as[size - 1] + bs[size - 1] < a + b) {
      insert(a, b);
    }
  }

  public void update(long a) {
    if (size == 0) {
      return;
    }
    as[size - 1] = Math.min(as[size - 1], a);
    while (size > 1) {
      pushDown();
      long e = event(as[size - 2], bs[size - 2], bs[size - 1]);
      if (e >= as[size - 1]) {
        pop();
      } else {
        break;
      }
    }
  }

  public long query() {
    return as[size - 1] + bs[size - 1];
  }

  public boolean empty() {
    return size == 0;
  }
}

File: 24_高级数据结构/good java/MinPairContainer/MinPairContainer.go
// 小偷问题
// 维护两个序列 a 和 b，支持以下操作：
// 1. 插入新元素 a[i] 和 b[i]（a[i] 应当是 a 中的最大元素）
// 2. 给定 x，对所有 i，将 a[i] 设为 min(a[i], x)
// 3. 查询 max(a[i]+b[i])
// 每个操作的均摊时间复杂度为 O(1).

package main

func main() {
	C := NewMinPairContainer(10)
	C.Add(1, 2)
	C.Add(2, 3)
	C.Add(3, 4)
	C.Update(2)
}

const INF int = 4e18

type MinPairContainer struct {
	as, bs []int
	size   int32
}

func NewMinPairContainer(n int) *MinPairContainer {
	return &MinPairContainer{
		as: make([]int, n),
		bs: make([]int, n),
	}
}

func (mpc *MinPairContainer) Add(a, b int) {
	for mpc.size > 1 {
		mpc._pushDown()
		e1 := mpc._event(mpc.as[mpc.size-2], mpc.bs[mpc.size-2], mpc.bs[mpc.size-1])
		e2 := mpc._event(mpc.as[mpc.size-1], mpc.bs[mpc.size-1], b)
		if e1 >= e2 {
			mpc._pop()
		} else {
			break
		}
	}
	if mpc.size == 0 || mpc.as[mpc.size-1]+mpc.bs[mpc.size-1] < a+b {
		mpc._insert(a, b)
	}
}

func (mpc *MinPairContainer) Update(a int) {
	if mpc.size == 0 {
		return
	}
	mpc.as[mpc.size-1] = min(mpc.as[mpc.size-1], a)
	for mpc.size > 1 {
		mpc._pushDown()
		e := mpc._event(mpc.as[mpc.size-2], mpc.bs[mpc.size-2], mpc.bs[mpc.size-1])
		if e >= mpc.as[mpc.size-1] {
			mpc._pop()
		} else {
			break
		}
	}
}

func (mpc *MinPairContainer) Query() int {
	return mpc.as[mpc.size-1] + mpc.bs[mpc.size-1]
}

func (mpc *MinPairContainer) Empty() bool {
	return mpc.size == 0
}

func (mpc *MinPairContainer) Clear() {
	mpc.size = 0
}

func (mpc *MinPairContainer) _event(a0, b0, b1 int) int {
	if b0 <= b1 {
		return -INF
	}
	return a0 + b0 - b1
}

func (mpc *MinPairContainer) _insert(a, b int) {
	mpc.as[mpc.size] = a
	mpc.bs[mpc.size] = b
	mpc.size++
}

func (mpc *MinPairContainer) _pop() {
	mpc.size--
}

func (mpc *MinPairContainer) _pushDown() {
	mpc.as[mpc.size-2] = min(mpc.as[mpc.size-2], mpc.as[mpc.size-1])
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
